package com.cilazatta.vollMed.exceptions;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;

import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import jakarta.persistence.EntityNotFoundException;

@RestControllerAdvice
public class TratadorDeErros {

	@ExceptionHandler({EntityNotFoundException.class, NoSuchElementException.class} )
	public ResponseEntity  tratarErro404() {
		return ResponseEntity.notFound().build();
	}
	

	@ExceptionHandler(BadCredentialsException.class)
	public ResponseEntity tratarErro403(BadCredentialsException ex) {
		var erros = ex.getMessage();
		return ResponseEntity.status(403).body(erros);
	}
	
	
	
	@ExceptionHandler(MethodArgumentNotValidException.class)
	public ResponseEntity tratarErro400(MethodArgumentNotValidException ex) {
		var erros = ex.getFieldErrors();
		List<DadosErroValidacaoDTO> listErros = erros.stream()
				.map(DadosErroValidacaoDTO::new)
				.collect(Collectors.toList());
		return ResponseEntity.badRequest().body(listErros);
	}

	
	
	
//===================================================================	
	private record DadosErroValidacaoDTO(String campo, String mensagem) {
		public DadosErroValidacaoDTO(FieldError erro) {
			this(erro.getField(), erro.getDefaultMessage());
		}
	}
	
}
